timesync/clock.go

```diff
@@ -59,39 +59,35 @@ func (t *TimeClock) startClock() error {
 	t.log.Info("starting global clock now=%v genesis=%v %p", t.clock.Now(), t.genesis, t)
 
 	for {
-		currLayer := t.Ticker.TimeToLayer(t.clock.Now()) // get current layer
+		currLayer := t.Ticker.TimeToLayer(t.clock.Now())
 		nextLayer := currLayer.Add(1)
 		if time.Until(t.Ticker.LayerToTime(currLayer)) > 0 {
 			nextLayer = currLayer
 		}
-		nextTickTime := t.Ticker.LayerToTime(nextLayer) // get next tick time for the next layer
-		diff := nextTickTime.Sub(t.clock.Now())
+		nextTickTime := t.Ticker.LayerToTime(nextLayer)
 		t.log.With().Info("global clock going to sleep before next layer",
-			log.String("diff", diff.String()),
 			log.FieldNamed("curr_layer", currLayer),
-			log.FieldNamed("next_layer", nextLayer))
-		tmr := time.NewTimer(diff)
-		select {
-		case <-tmr.C:
-			tmr.Stop()
-			t.mu.Lock()
-			subscriberCount := len(t.subscribers)
-			t.mu.Unlock()
-
-			t.log.With().Info("clock notifying subscribers of new layer tick",
-				log.Int("subscriber_count", subscriberCount),
-				t.TimeToLayer(t.clock.Now()))
-
-			// notify subscribers
-			if missed, err := t.Notify(); err != nil {
-				t.log.With().Warning("could not notify all subscribers",
-					log.Err(err),
-					log.Int("missed", missed))
+			log.FieldNamed("next_layer", nextLayer),
+			log.Time("next_tick_time", nextTickTime),
+		)
+
+		for {
+			// `time.After` sometimes unblocks bit too soon. In this case - wait again.
+			// See https://github.com/spacemeshos/go-spacemesh/issues/3617.
```

Comment:
this comment seems inconsistent with the issue. the issue is about the timer, not `time.After`.
are they the same thing?

if not, why not wrap the original code with this for loop with if check so the issue is on point?

Reply:
`time.After` and `time.Timer` share the same issue. The mentions `time.After` because this is used in the code right now.

I changed `time.Timer` to `time.After` because the code is simpler - no need to remember to `.Stop()` it manually. Is there any benefit of using `time.Timer` that I'm not aware of?

Reply:
```
// After waits for the duration to elapse and then sends the current time
// on the returned channel.
// It is equivalent to NewTimer(d).C.
// The underlying Timer is not recovered by the garbage collector
// until the timer fires. If efficiency is a concern, use NewTimer
// instead and call Timer.Stop if the timer is no longer needed.
func After(d Duration) <-chan Time {
	return NewTimer(d).C
}
```

Reply:
Thanks, I wasn't aware :)